workitem/workitem_repository.go

```diff
@@ -701,50 +702,201 @@ func (r *GormWorkItemRepository) Fetch(ctx context.Context, spaceID uuid.UUID, c
+// GetCountsPerIteration counts WIs including iteration-children and returns a map of iterationID->WICountsPerIteration
+func (r *GormWorkItemRepository) GetCountsPerIteration(ctx context.Context, spaceID uuid.UUID) (map[string]WICountsPerIteration, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "workitem", "getCountsPerIteration"}, time.Now())
+	db := r.db.Model(&iteration.Iteration{}).Where("space_id = ?", spaceID)
+	if db.Error != nil {
+		return nil, errors.NewInternalError(ctx, db.Error)
+	}
+
+	wiMap, err := r.getAllIterationWithCounts(ctx, db, spaceID)
+	if err != nil {
+		return nil, err
+	}
+
+	countsMap, err := r.getFinalCountAddingChild(ctx, db, spaceID, wiMap)
+	if err != nil {
+		return nil, err
+	}
+	return countsMap, nil
+}
+
+// GetCountsForIteration returns Closed and Total counts of WIs for given iteration
+// It fetches all child iterations of input iteration and then uses list to counts work items
+// SELECT count(*) AS Total,
+//        count(CASE fields->>'system.state'
+//                  WHEN 'closed' THEN '1'
+//                  ELSE NULL
+//              END) AS Closed
+// FROM work_items wi
+// WHERE fields->>'system.iteration' IN ('input iteration ID + children IDs')
+//   AND wi.deleted_at IS NULL
+func (r *GormWorkItemRepository) GetCountsForIteration(ctx context.Context, itr *iteration.Iteration) (map[string]WICountsPerIteration, error) {
+	defer goa.MeasureSince([]string{"goa", "db", "workitem", "getCountsForIteration"}, time.Now())
 	var res WICountsPerIteration
-	query := fmt.Sprintf(`SELECT count(*) as Total,
-						count( case fields->>'system.state' when 'closed' then '1' else null end ) as Closed
-						FROM "work_items"
-						where fields@> concat('{"system.iteration": "%s"}')::jsonb
-						and work_items.deleted_at is null`, iterationID)
-	db := r.db.Raw(query)
+	pathOfIteration := append(itr.Path, itr.ID)
+	// get child IDs of the iteration
+	var childIDs []uuid.UUID
+	iterationTable := iteration.Iteration{}
+	iterationTableName := iterationTable.TableName()
+	getIterationsOfSpace := fmt.Sprintf(`SELECT id FROM %s WHERE path <@ ? and space_id = ?`, iterationTableName)
+	db := r.db.Raw(getIterationsOfSpace, pathOfIteration.Convert(), itr.SpaceID.String())
+	db.Pluck("id", &childIDs)
+	if db.Error != nil {
+		log.Error(ctx, map[string]interface{}{
+			"path": pathOfIteration.Convert(),
+			"err":  db.Error,
+		}, "unable to fetch children for path")
+		return nil, errors.NewInternalError(ctx, db.Error)
+	}
+	childIDs = append(childIDs, itr.ID)
+
+	// build where clause usig above ID list
+	idsToLookFor := []string{}
+	for _, x := range childIDs {
+		idsToLookFor = append(idsToLookFor, x.String())
+	}
+	whereClause := strings.Join(idsToLookFor, ",")
+	query := fmt.Sprintf(`SELECT count(*) AS Total,
+						count(CASE fields->>'system.state'
+									WHEN 'closed' THEN '1'
+									ELSE NULL
+								END) AS Closed
+					FROM %s wi
+					WHERE fields->>'system.iteration' IN (?)
```

Comment:
using GIN index now https://github.com/almighty/almighty-core/pull/1151/commits/095d55560a16f8aa82afbba4e87d27bfff8726a3
thanks.
Query string got longer + not much readable but it is now using `@>` for `WHERE IN` clause